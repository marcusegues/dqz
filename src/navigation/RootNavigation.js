import React from 'react';
import { Notifications } from 'expo';
import { translate } from 'react-i18next';
import { StackNavigator } from 'react-navigation';
import i18n from '../i18n';
import OnBoarding from '../screens/OnBoarding/OnBoarding';
import ScreensView from '../screens/ScreensView/ScreensView';
import { MAIN_RED, MAIN_BACKGROUND_COLOR } from '../styles/colors';
import { defaultNavigationOptions } from './navigationOptions';
import HeaderTitle from '../components/Headers/subcomponents/HeaderTitle';
import Logo from '../components/Headers/subcomponents/Logo';
import OptionsButton from '../components/Headers/subcomponents/OptionsButton';
import InfoIcon from '../components/Headers/subcomponents/InfoIcon';
import MainMenuHeaderRight from '../components/Headers/subcomponents/MainMenuHeaderRight';
import MainMenu from '../screens/MainMenu/MainMenu';
import GoodQuantityInput from '../components/QuestionAnswer/QuantityInput/GoodQuantityListModal';

import registerForPushNotificationsAsync from '../../api/registerForPushNotificationsAsync';
import { UnderConstruction } from './underConstruction';
import QuestionAnswerContainer from '../components/QuestionAnswer/QuestionAnswerContainer';
import UnderConstructionWithSums from './underConstructionWithSums';

const RootStackNavigator = StackNavigator(
  {
    Screens: {
      screen: ScreensView,
      navigationOptions: ({ navigationOptions }) => ({
        ...navigationOptions,
        headerTitle: <HeaderTitle text="Screens" />,
        headerLeft: <Logo />,
        headerRight: <OptionsButton />,
        headerStyle: {
          ...navigationOptions.headerStyle,
          borderBottomWidth: 5,
          borderBottomColor: MAIN_RED,
        },
      }),
    },
    QuestionAnswer: {
      screen: QuestionAnswerContainer,
      navigationOptions: () => ({
        headerTitle: <HeaderTitle text="Waren deklarieren - 2" />,
      }),
    },
    OnBoarding: {
      screen: OnBoarding,
      navigationOptions: () => ({
        header: null,
      }),
    },
    GoodQuantityInput: {
      screen: GoodQuantityInput,
    },
    MainMenu: {
      screen: MainMenu,
      navigationOptions: ({ navigationOptions, navigation }) => ({
        ...navigationOptions,
        headerLeft: <InfoIcon navigation={navigation} />,
        headerRight: <MainMenuHeaderRight navigation={navigation} />,
        headerStyle: {
          ...navigationOptions.headerStyle,
          elevation: 0,
        },
      }),
    },

    UnderConstruction: {
      screen: UnderConstruction,
      navigationOptions: () => ({
        headerTitle: <HeaderTitle text="In Bearbeitung" />,
      }),
    },
    UnderConstructionWithSums: {
      screen: UnderConstructionWithSums,
      navigationOptions: () => ({
        headerTitle: <HeaderTitle text="In Bearbeitung" />,
      }),
    },
  },
  {
    navigationOptions: defaultNavigationOptions,
    cardStyle: { backgroundColor: MAIN_BACKGROUND_COLOR },
    initialRouteName: 'QuestionAnswer',
  }
);

const WrappedRootStackNavigator = () => (
  <RootStackNavigator screenProps={{ t: i18n.getFixedT() }} />
);

const ReloadAppOnLanguageChange = translate(null, {
  bindI18n: 'languageChanged',
  bindStore: false,
})(WrappedRootStackNavigator);

export default class RootNavigator extends React.Component {
  componentDidMount() {
    this._notificationSubscription = this._registerForPushNotifications();
  }

  componentWillUnmount() {
    // eslint-disable-next-line
    this._notificationSubscription && this._notificationSubscription.remove();
  }

  _registerForPushNotifications() {
    // Send our push token over to our backend so we can receive notifications
    // You can comment the following line out if you want to stop receiving
    // a notification every time you open the app. Check out the source
    // for this function in api/registerForPushNotificationsAsync.js
    registerForPushNotificationsAsync();

    // Watch for incoming notifications
    this._notificationSubscription = Notifications.addListener(
      this._handleNotification
    );
  }

  _handleNotification = () => {};

  render() {
    return <ReloadAppOnLanguageChange />;
  }
}
