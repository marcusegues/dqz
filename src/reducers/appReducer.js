/* eslint-disable prefer-destructuring */
// @flow
import Immutable from 'immutable';
import {
  getInitialState,
  mainCategories,
  EmptyMainCategories,
} from '../types/reducers/appReducer';
import type {
  State,
  MainCategory,
  MainCategories,
  Settings,
} from '../types/reducers/appReducer';
import type { Action } from '../types/actions';
import type {
  Amounts,
  Basket,
  Category,
  People,
} from '../model/types/basketPeopleAmountsTypes';
import * as modelApi from '../model/configurationApi';
import type { Currency, CurrencyObject } from '../model/currencies';
import type { PaymentData } from '../types/generalTypes';

export const appState = (
  state: State = getInitialState(),
  action: Action
): State => {
  switch (action.type) {
    case 'UPDATE_CURRENCIES': {
      const currencyObject: CurrencyObject = action.currencyObject;
      const validCurrencies: boolean = action.validCurrencies;
      const currencyDate: Date = action.currencyDate;

      const s1 = state.set('currencyObject', currencyObject);
      const s2 = s1.set('currencyDate', currencyDate);
      return s2.set('validCurrencies', validCurrencies);
    }

    case 'BASKET_ADD_QUANTITY': {
      const category: Category = action.category;
      const date: string = action.date;
      const basket: Basket = state.get('basket');
      return state.setIn(
        ['basket'],
        modelApi.addQuantity(
          basket,
          category,
          action.quantity,
          date.length > 0 ? date : undefined
        )
      );
    }
    case 'SET_BASKET': {
      const basket: Basket = action.basket;
      return state.set('basket', basket);
    }

    case 'ADULTS_CHANGE_QUANTITY': {
      const people: People = state.get('people');
      const adults: number = people.get('adults');
      const quantity: number = adults + action.quantityChange;
      return state.set('people', modelApi.setAdultPeople(people, quantity));
    }
    case 'MINORS_CHANGE_QUANTITY': {
      const people: People = state.get('people');
      const minors: number = people.get('minors');
      const quantity: number = minors + action.quantityChange;
      return state.set('people', modelApi.setMinorPeople(people, quantity));
    }
    case 'ADULTS_SET_QUANTITY': {
      const people: People = state.get('people');
      const quantity: number = action.quantity;
      return state.set('people', modelApi.setAdultPeople(people, quantity));
    }
    case 'MINORS_SET_QUANTITY': {
      const people: People = state.get('people');
      const quantity: number = action.quantity;
      return state.set('people', modelApi.setMinorPeople(people, quantity));
    }
    case 'SET_PEOPLE': {
      return state.set('people', action.people);
    }

    case 'SET_AMOUNTS': {
      const amounts: Amounts = action.amounts;
      return state.set('amounts', amounts);
    }
    case 'ADD_AMOUNT': {
      const currency: Currency = action.currency;
      const amounts: Amounts = state.get('amounts');
      const amount: number = action.amount;
      return state.set(
        'amounts',
        modelApi.addAmount(amounts, currency, amount)
      );
    }
    case 'ADD_LARGE_AMOUNT': {
      const currency: Currency = action.currency;
      const largeAmount: number = action.largeAmount;
      const amounts: Amounts = state.get('amounts');
      return state.set(
        'amounts',
        modelApi.addLargeAmount(amounts, currency, largeAmount)
      );
    }
    case 'RESET_LARGE_AMOUNTS': {
      const amounts: Amounts = state.get('amounts');
      const currency: Currency = action.currency;
      return state.set(
        'amounts',
        modelApi.resetLargeAmounts(amounts, currency)
      );
    }
    case 'RESET_AMOUNTS': {
      const amounts: Amounts = state.get('amounts');
      const currency: Currency = action.currency;
      return state.set('amounts', modelApi.resetAmounts(amounts, currency));
    }

    case 'RESET_DECLARATION': {
      const peopleReset = state.set('people', modelApi.initPeople);
      const basketReset = peopleReset.set('basket', modelApi.emptyBasket);
      const mainCategoriesReset = basketReset.setIn(
        ['settings', 'mainCategories'],
        Immutable.Set()
      );
      return mainCategoriesReset.set('amounts', modelApi.initAmounts);
    }

    case 'ADD_MAIN_CATEGORY': {
      const mainCategory: MainCategory = action.mainCategory;
      const mainCategoriesAnswer = state.getIn(
        ['settings', 'mainCategories'],
        EmptyMainCategories
      );
      return state.setIn(
        ['settings', 'mainCategories'],
        mainCategoriesAnswer.add(mainCategory)
      );
    }
    case 'REMOVE_MAIN_CATEGORY': {
      const mainCategory: MainCategory = action.mainCategory;
      const mainCategoriesAnswer = state.getIn(
        ['settings', 'mainCategories'],
        EmptyMainCategories
      );
      return state.setIn(
        ['settings', 'mainCategories'],
        mainCategoriesAnswer.delete(mainCategory)
      );
    }
    case 'SET_MAIN_CATEGORIES': {
      const mainCategoriesAnswer: MainCategories = action.mainCategories;
      return state.setIn(['settings', 'mainCategories'], mainCategoriesAnswer);
    }

    case 'SET_PAYMENT_DATA': {
      const paymentData: PaymentData = action.paymentData;
      return state.set('paymentData', paymentData);
    }

    case 'SET_RECEIPT_ID': {
      const receiptId: string = action.receiptId;
      return state.set('receiptId', receiptId);
    }

    case 'SET_RECEIPT_ENTRY_TIME': {
      const receiptEntryTime: string = action.receiptEntryTime;
      return state.set('receiptEntryTime', receiptEntryTime);
    }

    default: {
      return state;
    }
  }
};

export const getBasket = (state: State): Basket => state.get('basket');

export const getPeople = (state: State): People => state.get('people');

export const getAmounts = (state: State): Amounts => state.get('amounts');

export const getMainCategories = (state: State): MainCategories =>
  state.getIn(['settings', 'mainCategories'], mainCategories);

export const getSettings = (state: State): Settings => state.get('settings');

export const getCurrenciesObject = (state: State): CurrencyObject =>
  state.get('currencyObject');

export const getCurrencyState = (state: State): boolean =>
  state.get('validCurrencies');

export const getCurrencyDate = (state: State): Date =>
  state.get('currencyDate');

export const getPaymentData = (state: State): PaymentData =>
  state.get('paymentData');

export const getReceiptId = (state: State): string => state.get('receiptId');

export const getReceiptEntryTime = (state: State): string =>
  state.get('receiptEntryTime');
