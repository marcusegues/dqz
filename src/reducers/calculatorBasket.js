import { emptyBasket } from '../constants/basket';

const calculatorBasket = (state = emptyBasket, action) => {
  switch (action.type) {
    case 'CHANGE_QUANTITY_CALCULATOR_BASKET_ITEM': {
      return {
        ...state,
        [action.itemName]: state[action.itemName] + action.quantityChange,
      };
    }
    default: {
      return state;
    }
  }
};

export default calculatorBasket;
