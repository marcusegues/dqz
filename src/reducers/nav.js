import { NavigationActions } from 'react-navigation';
import { RootStackNavigator } from '../navigation/RootNavigation';

export const initialNavState = RootStackNavigator.router.getStateForAction(
  RootStackNavigator.router.getActionForPathAndParams('OnBoarding')
);

export const nav = (state = initialNavState, action) => {
  let navigationAction;
  switch (action.type) {
    case 'NAVIGATE': {
      navigationAction = NavigationActions.navigate({
        routeName: action.screen,
        params: action.params,
      });
      break;
    }
    case 'SET_PARAMS': {
      navigationAction = NavigationActions.setParams({
        params: action.params,
        key: action.key || state.routes[state.index].key,
      });
      break;
    }
    case 'GO_BACK': {
      navigationAction = NavigationActions.back({
        key: action.key || null,
      });
      break;
    }
    default: {
      return state;
    }
  }
  const nextState = RootStackNavigator.router.getStateForAction(
    navigationAction,
    state
  );
  return nextState;
};
