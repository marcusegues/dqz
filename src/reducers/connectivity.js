import type { State } from '../types/reducers/appReducer';
import { getInitialState } from '../types/reducers/connectivityReducer';
import type { ConnectivityState } from '../types/reducers/connectivityReducer';
import type { ConnectivityAction } from '../types/actions/connectivity';

export const connectivity = (
  state: ConnectivityState = getInitialState(),
  action: ConnectivityAction
): State => {
  switch (action.type) {
    case 'SET_CONNECTIVITY': {
      return action.connectionInfo;
    }
    default: {
      return state;
    }
  }
};
