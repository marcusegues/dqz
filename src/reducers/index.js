// @flow
import { combineReducers } from 'redux';
import * as appApi from './appReducer';
import * as connectivityApi from './connectivity';
import { connectivity } from './connectivity';
import { formatDate } from '../model/utils';
import type {
  Amounts,
  Basket,
  People,
} from '../model/types/basketPeopleAmountsTypes';
import type {
  MainCategories,
  Settings,
  State,
} from '../types/reducers/appReducer';
import type { CurrencyObject } from '../model/currencies';
import { calculateVat } from '../model/vatCalculations';
import { calculateDuty } from '../model/dutyCalculations';
import type { DutyReport, VatReport } from '../model/types/calculationTypes';
import type { PaymentData } from '../types/generalTypes';
import type { ConnectivityType } from '../types/connectivity';
import type { AppState } from '../types/reducers';

/**
 * Combines reducers (currently just one)
 * @type {Reducer<any>}
 */
export const root: AppState = combineReducers({
  appState: appApi.appState,
  connectivity,
});

/**
 * Gets basket in application state
 * @param state
 * @returns {Basket}
 */
export const getBasket = (state: AppState): Basket =>
  appApi.getBasket(state.appState);

/**
 * Gets people in application state
 * @param state
 * @returns {People}
 */
export const getPeople = (state: AppState): People =>
  appApi.getPeople(state.appState);

/**
 * Gets amounts in application state
 * @param state
 * @returns {Amounts}
 */
export const getAmounts = (state: AppState): Amounts =>
  appApi.getAmounts(state.appState);

/**
 * Gets main categories in application state
 * @param state
 * @returns {MainCategories}
 */
export const getMainCategories = (state: AppState): MainCategories =>
  appApi.getMainCategories(state.appState);

/**
 * Gets settings in application state
 * @param state
 * @returns {Settings}
 */
export const getSettings = (state: AppState): Settings =>
  appApi.getSettings(state.appState);

/**
 * Gets currencies in application state
 * @param state
 * @returns {CurrencyObject}
 */
export const getCurrencies = (state: AppState): CurrencyObject =>
  appApi.getCurrenciesObject(state.appState);

/**
 * Gets currency state in application state
 * @param state
 * @returns {boolean}
 */
export const getCurrencyState = (state: AppState): boolean =>
  appApi.getCurrencyState(state.appState);

/**
 * Gets the formatted currency date in application state
 * @param state
 * @returns {string}
 */
export const getFormattedCurrencyDate = (state: AppState): string =>
  formatDate(appApi.getCurrencyDate(state.appState));

/**
 * Gets the total vat in application state
 * @param state
 */
export const getTotalVat = (state: AppState): number =>
  calculateVat(
    state.appState.amounts,
    state.appState.people,
    state.appState.currencyObject
  ).get('totalVat', 0);

/**
 * Gets the total duty in application state
 * @param state
 */
export const getTotalDuty = (state: AppState): number =>
  calculateDuty(state.appState.basket, state.appState.people).get(
    'totalDuty',
    0
  );

/**
 * Gets the total fees (== vat + duty) in application state
 * @param state
 * @returns {number}
 */
export const getTotalFees = (state: AppState): number =>
  getTotalVat(state) + getTotalDuty(state);

/**
 * Gets the duty report in application state
 * @param state
 * @returns {DutyReport}
 */
export const getDutyReport = (state: AppState): DutyReport =>
  calculateDuty(state.appState.basket, state.appState.people);

/**
 * Gets the vat report in application state
 * @param state
 * @returns {VatReport}
 */
export const getVatReport = (state: AppState): VatReport =>
  calculateVat(
    state.appState.amounts,
    state.appState.people,
    state.appState.currencyObject
  );

/**
 * Gets current payment data
 * @param state
 * @returns {PaymentData}
 */
export const getPaymentData = (state: AppState): PaymentData =>
  appApi.getPaymentData(state.appState);

/**
 * Gets the current receipt id (emptystring if not active)
 * @param state
 * @returns {string}
 */
export const getReceiptId = (state: AppState): string =>
  appApi.getReceiptId(state.appState);

/**
 * Gets the current receipt entry time (empty string if not set yet)
 * @param state
 * @returns {string}
 */
export const getReceiptEntryTime = (state: AppState): string =>
  appApi.getReceiptEntryTime(state.appState);

/**
 * Gets the current receipt id (emptystring if not active)
 * @param state
 * @returns {string}
 */
export const getConnectivity = (state: AppState): ConnectivityType =>
  connectivityApi.getConnectivity(state.connectivity);
