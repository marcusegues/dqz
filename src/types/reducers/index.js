// @flow
import type { RecordOf, RecordFactory } from 'immutable';
import Immutable from 'immutable';
import { getInitialDeclarationState } from './declaration';
import { getInitialConnectivityState } from './connectivity';
import type { DeclarationState } from './declaration';
import type { ConnectivityState } from './connectivity';
import type { CurrenciesState } from './currencies';
import type { ReceiptsState } from './receipts';
import { getInitialReceiptsState } from './receipts';
import { getInitialCurrenciesState } from './currencies';
import type { NavState } from './nav';
import {getInitialNavState} from "./nav";

type AppStateObj = {
  declaration: DeclarationState,
  connectivity: ConnectivityState,
  currencies: CurrenciesState,
  receipts: ReceiptsState,
  nav: NavState,
};

export const getInitialAppState: RecordFactory<AppStateObj> = Immutable.Record({
  declaration: getInitialDeclarationState(),
  connectivity: getInitialConnectivityState(),
  currencies: getInitialCurrenciesState(),
  receipts: getInitialReceiptsState(),
  nav: getInitialNavState(),
});

export type AppState = RecordOf<AppStateObj>;
