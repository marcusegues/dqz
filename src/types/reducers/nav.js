// @flow
import type { RecordOf, RecordFactory } from 'immutable';
import Immutable from 'immutable';
import type { ConnectivityType } from '../connectivity';
import { initConnectivity } from '../connectivity';
import { initialNavState } from '../../reducers/nav';

type NavStateObj = {
  index: number,
  routes: Object,
};

export const getInitialNavState: RecordFactory<NavStateObj> = Immutable.Record({
  ...initialNavState,
});

export type NavState = RecordOf<NavStateObj>;
