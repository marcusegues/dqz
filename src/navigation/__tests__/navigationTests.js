// @flow

import { stackNavigatorScreens } from '../RootNavigation';
import { nav } from '../../reducers/nav';

describe('StackNavigation', () => {
  test('has OnBoarding as initial route', () => {
    expect(nav(undefined, { type: 'INIT' }).routes[0].routeName).toBe(
      'OnBoarding'
    );
  });

  test('has all screens', () => {
    expect(Object.keys(stackNavigatorScreens).sort()).toMatchSnapshot();
  });
});
