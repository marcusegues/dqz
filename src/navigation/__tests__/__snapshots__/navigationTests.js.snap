// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackNavigation has all screens 1`] = `
Array [
  "CurrencyPickerModal",
  "GoodQuantityListModal",
  "MainMenu",
  "OnBoarding",
  "Payment",
  "QuestionAnswer",
  "Screens",
  "UnderConstruction",
  "UnderConstructionWithSums",
]
`;
